processing/src/main/java/org/apache/druid/segment/join/HashJoinSegmentStorageAdapter.java

```diff
@@ -209,13 +209,36 @@ public Metadata getMetadata()
       @Nullable final QueryMetrics<?> queryMetrics
   )
   {
-    if (!Objects.equals(joinFilterPreAnalysis.getOriginalFilter(), filter)) {
-      throw new ISE(
-          "Filter provided to cursor [%s] does not match join pre-analysis filter [%s]",
+    JoinFilterPreAnalysis jfpa;
+    if (filter == null) {
+      jfpa = JoinFilterAnalyzer.computeJoinFilterPreAnalysis(
```

Comment:
Does this mean we'll re-compute the pre-analysis every time if the filter is null?

I think this code path is short-circuited right now, but it could change in the future and it would be hard to remember that the computation is not cached.

Reply:
I adjusted the analysis group so the key is now (Filter, JoinableClauses, VirtualColumns) instead of filter, I think that's a more correct key, and the null filters are cached now.